Implement undoing moves and moving backwards

ubeF opened the pull request.

ubeF committed:
Implemented Resetting Moves and Moving Backwards

hfxbse commented on map/map.gd:
`PlayerLocation` should be in a separate file to not pollute too much.

The method and function names should be `snake_case` like the rest.

hfxbse reviewed (changes requested):
I am still not sure if keeping all data regarding the player position for each player inside the map itself. The code it requires to track the relationship between the player and their location adds a lot of complexity. Keeping that information inside the player class feels more natural to me, as `PlayerLocation` already has "player" in it.

This would simplify the map to only a scene of fields and a reference to the start field. Considering fields can influence the movement heavily, like ending a turn early, it might be the right call in the long run.

Small details optical nature, undoing the previous move can be a teleport no issue, but moving backward without an animation feels weird :)

Undoing twice does not seem possible, I would opt to undo the undoing. Might not be necessary depending on the details of the game, I would have to look that up myself.

ubeF commented:
> I am still not sure if keeping all data regarding the player position for each player inside the map itself. The code it requires to track the relationship between the player and their location adds a lot of complexity. Keeping that information inside the player class feels more natural to me, as `PlayerLocation` already has "player" in it.

Putting the logic and data for resetting and moving backwards into the player object would split the move "interface", with move being defined on Map and moveBackwards defined on Player. Keeping the data in the Player object and the logic in Map would introduce an unnecessary amount of coupling. Moving both the data and the logic into Map like i did here seems like the best option IMO

> Small details optical nature, undoing the previous move can be a teleport no issue, but moving backward without an animation feels weird :)

Not sure on how to fix this and if it's even worth the time at this point

hfxbse commented:
> Putting the logic and data for resetting and moving backwards into the player object would split the move "interface", with move being defined on Map and moveBackwards defined on Player. Keeping the data in the Player object and the logic in Map would introduce an unnecessary amount of coupling. Moving both the data and the logic into Map like i did here seems like the best option IMO

Yeah obviously don't split the movement interface between objects, but I believe it is more reasonable to move the movement into the player, removing the need to track the location and player relation. The map really does not do anything the player could not do itself.

> > Small details optical nature, undoing the previous move can be a teleport no issue, but moving backward without an animation feels weird :)
> 
> Not sure on how to fix this and if it's even worth the time at this point

Adding a move backward method to the path should be enough, which should be nearly identical to the move forward method. But feel free to priorities something else if you prefer.

ubeF commented:
> > Putting the logic and data for resetting and moving backwards into the player object would split the move "interface", with move being defined on Map and moveBackwards defined on Player. Keeping the data in the Player object and the logic in Map would introduce an unnecessary amount of coupling. Moving both the data and the logic into Map like i did here seems like the best option IMO
> 
> Yeah obviously don't split the movement interface between objects, but I believe it is more reasonable to move the movement into the player, removing the need to track the relation between players and their relation. The map really does not do anything the player could not do itself.
> 
> > > Small details optical nature, undoing the previous move can be a teleport no issue, but moving backward without an animation feels weird :)
> > 
> > 
> > Not sure on how to fix this and if it's even worth the time at this point
> 
> Adding a move backward method to the path should be enough, which should be nearly identical to the move forward method. But feel free to prioritize something else if you prefer.

Will try that

ubeF committed:
Transferred movement-logic from map to player

ubeF committed:
Changed function-names to snake-case

ubeF commented:
Adding animation for moving backwards turns out to be more complicated than expected. I'd prefer merging this branch for now and addressing the animation later

hfxbse committed:
use const for constant value instead of var

hfxbse committed:
change remainging class memebers to snake casing

hfxbse committed:
change wording from reset to undo

Reset implies an initial state, while undo does not. When using
reset, it should be "reset TO last move", which is longer the
undo counterpart.
